problemtools/verifyproblem.py

```diff
@@ -947,8 +994,47 @@ def check(self, args):
                     self.warning('%s gets AC' % (desc))
             os.unlink(file_name)
 
+        self._verify_invalid_outputs()
+
         return self._check_res
 
+    def _verify_invalid_outputs(self):
+        """Check that output validators decline invalid answer files"""
+
+        val_timelim = self._problem.config.get('limits')['validation_time']
+        val_memlim = self._problem.config.get('limits')['validation_memory']
+        flags = self._problem.config.get('validator_flags').split()
+
+        for testcase in self._problem.testdata.get_all_testcases():
+            if not os.path.exists(testcase.infile):
+                # .in-file doesn't exist, already reported by testcase-check, so 'ignore' here
+                continue
+
+            wrong_answers = glob.glob(OutputValidators.WA_GLOB % testcase.ansfile)
+            if wrong_answers:
+                if self._problem.is_interactive:
+                    self.warning('Output validator check with wrong answers for interactive problems is currently not supported (skipping)')
+                    return
+
+                if self._problem.config.get('validation') == 'default':
+                    self.warning('Output validator check with wrong answers and default validator is unnecessary (skipping)')
+                    return
+
+            for wafile in wrong_answers:
+                for val in self._validators:
+                    feedbackdir = tempfile.mkdtemp(prefix='feedback', dir=self._problem.tmpdir)
+                    status, runtime = val.run(wafile,
+                                              args=[testcase.infile, testcase.ansfile, feedbackdir] + flags,
+                                              timelim=val_timelim, memlim=val_memlim)
+                    res = self._parse_validator_results(val, status, feedbackdir, testcase)
+                    shutil.rmtree(feedbackdir)
+                    if res.verdict != 'AC':
+                        # One output validator declined this wrong answer, so stop validating
+                        break
+                else:
+                    # Will only be executed if loop wasn't ended by break
+                    # No output validator declined wrong answer, this is an error.
+                    self.error("Wrong answer file %s was not declined by output validators" % wafile)
```

Comment:
Like with the input validator, I don't like this duplication much (pretty complex code, too).

Reply:
I think a function here would be nice. Maybe an abstract base class 'Validators' would help here, this would also make the check on noncompiling validators easier.

Reply:
To abstract both input and output validators? I'm not sure, they don't share that much logic, and inheritance tends to be confusing. Maybe a helper function could be better (if anything; some small amount of duplication is fine IMO).